### ISSUE TITLE
How to conceptualize subtyping for legacy-nullness

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#33]
It's beyond debate that `@NotNull Foo <: @Nullable Foo`. When `@LegacyNull Foo` comes in, well, it should still be uncontroversial that assigning from not-null to legacy is safe, and that assigning from legacy to nullable is safe.

But then the questions of assigning legacy to not-null, and/or assigning nullable to legacy, are not as clear. I would say there are two choices of rule sets that both make good sense:

(1) The strict interpretation: neither of these cases is assignable; both are unsound.

(2) The lenient interpretation: both are assignable, because we don't have enough basis for complaint and we will annoy users.

Interestingly, (1) leads to a "normal" subtype arrangement of `not-null <: legacy <: nullable` where the anti-symmetric property is preserved. (2) leads to breaking not just anti-symmetry but also transitivity (we want `nullable <: legacy` and `legacy <: not-null` but we don't want `nullable <: not-null`!).

I think a clean solution may be this:  *assignability* rules really are defined conservatively as in (1). But we also have something akin to Java's "unchecked conversion". Even though `nullable </: legacy` and `legacy </: not-null`, you can still make that conversion anyway IF you `@SuppressWarnings("nullness:legacy")` (let's not debate the name though).

What's nice about this is that projects can decide whether they want to be strict or lenient based on whether they apply that warning suppression at a broad level or not. I think it's also much easier to think about this way, rather than saying only "`not-null <: nullable` but `legacy` is a special thing sitting off to the side somewhere".

**Comment by cpovirk** [comment]
[Way back when](https://github.com/jspecify/jspecify/issues/33#issuecomment-543360703), I wondered about the implications of circular subtyping.

I still don't have a great understanding of the question, but: From what I understand, there is only one sufficiently powerful approach to checking if one Java type is a subtype of another: by traversing the inheritance graph (whose edges connect a type to its direct supertypes, as defined in [JLS 4.10.2](https://docs.oracle.com/javase/specs/jls/se14/html/jls-4.html#jls-4.10.2), etc.).

I _believe_ that we, too, are stuck doing this even though we'd prefer to check "only the nullness part" of types. (To handwave: That's thanks to type variables, types that extend `Foo<@Nullable Bar>`, and probably other stuff.) I think someone (from JetBrains?) warned us that this would be the case a long time ago. I don't know that we heard details at the time, but I should give credit, if only I remembered who it was....

So, what does this mean? Well, first of all: If we're traversing a graph that has cycles in it, then we need to track which nodes we've already seen. We _might_ need to do this anyway, though??

We might also need to ensure that we complete any "definitely safe" traversals before we complete any "possibly safe" / "not enough information because of unspecified nullness" traversals. That way, we don't issue warnings spuriously.

This might happen naturally (especially?) if we use BFS over DFS, but (a) I'm not absolutely 100% confident in that and (b) BFS might complicate an already complicated implementation. (Another small complication: We need to _track_ whether we have used a "possibly safe" edge.)

I can think of another possible implementation, which is easier, albeit slightly wasteful: Run a full search of an inheritance graph _without_ the "possibly safe" edges first and then, if that fails, to run a search of an inheritance graph _with_ those edges.

There's good news for tools that operate only in "lenient mode": They don't need to worry about whether they use "possibly safe" edges or not, so they can run a single search with all edges present.

**Comment by cpovirk** [comment]
I still need to finish rereading this thread, but: Currently I'm less concerned with circular subtyping per se than I am with the idea that, _in the specific case of "lenient mode,"_ we might accidentally have `nullable <: unspecified <: non-null` (as kevinb9n notes in [the first post](https://github.com/jspecify/jspecify/issues/33#issue-458272713)).

One solution to this seems to be that `nullable` is never a _subtype_ of `unspecified`, but it can be ["consistent with"](https://link.springer.com/chapter/10.1007/978-3-319-89884-1_1) it. It's not clear to me whether people have extended this idea to parameterized types (and whatever other challenges we'd face in extending it to Java). I think I saw a paper that sounded related, but I haven't read it yet. But even if this isn't a fully solved problem, maybe it's easier to solve when the domain is only ~2 nullnesses instead of an arbitrary number of types?

Another solution is that subtyping of nullness is non-transitive. Non-transitive subtyping [is part of Kotlin](https://kotlin.github.io/kotlin-spec/#subtyping). (See also [flexible types](https://kotlin.github.io/kotlin-spec/#flexible-types) and [subtyping for flexible types](https://kotlin.github.io/kotlin-spec/#subtyping-for-flexible-types). Here are [Andrey's slides](https://www.dropbox.com/s/p5cqn74q7lqfpl3/Flexible%20Types%20in%20Kotlin%20-%20JVMLS%202015.pdf?dl=0) again. Probably I should also look at [subtyping for nullable types](https://kotlin.github.io/kotlin-spec/#subtyping-for-nullable-types).) I haven't looked into how they prevent `nullable <: non-null`. Maybe they refuse to follow the edge to `unspecified` unless they're *handwave* directly matching an `unspecified` type in the some known type that they're comparing to?

Hopefully I can understand this a little better and get to the point of having specific questions for Andrey / Denis / Werner / others.

**Comment by cpovirk** [comment]
I sent [an email to ask about "consistency"/"equivalence" as a way of supporting non-transitive subtyping](https://groups.google.com/d/topic/jspecify-dev/yPnkx_GSb0Q).

**Comment by kevinb9n** [comment]
I do like the "Many-Worlds Interpretation" (MWI) of unspecified nullness. What I think it says is that many seemingly-boolean propositions need instead to be answered with a ternary boolean {false, unknown, true}. Then we imagine forking off two parallel universes for each unspecified-nullness type in the system, then IF the question has the same true or false answer in ALL of those universes, we can definitively choose that as the real answer; otherwise the answer is "unknown".

(Or we just use the usual truth tables for ternary boolean logic; I still think this works out to the same thing.)

(Aside: when asking whether a type is nullable, it is helpful to standardize on that as the positive direction (i.e. mapping true=nullable, false=non-nullable); then we can speak of things like "intersecting" or "ANDing" nullness, and the nullness of an intersection type would be the intersection of the nullness of each component, etc. And it's conventional to view false < true as well as to view subtypes < supertypes... "it all just works out, man..")

Anyway, all this to say,

The part of the initial proposal here, that non-nullable is a subtype of unspecified and that unspecified is a subtype of nullable, is simply a consequence of MWI. Fork off two universes, get the same true answer in both, you're done.

But asking whether unspecified is a subtype of non-nullable yields the answer "unknown". This led to the idea that we would have a middle-of-the-road response: allow it, but with a warning. However, it sure feels like we need one definitive answer on whether these are subtypes or not. The initial proposal said "they're not, because subtyping is quite strict and also expected to be transitive and antisymmetric; but we can allow a type conversion with a warning anyway".

This last part seems to be the part that's remained the most controversial.

**Comment by cpovirk** [comment]
I support thinking of the null direction as "true." I like thinking of it as "Union with null? Yes," in contrast to "Union with null? No" and "Union with null? Unknown."

(I am glossing over some deeper questions, but we may be able to agree on this much without getting into those.)

I do think that we're going to need "unknown" to mean "yes" in lenient mode, even though that makes subtyping non-transitive. I'm encouraged that Kotlin gets away with [non-transitive subtyping](https://kotlin.github.io/kotlin-spec/#subtyping) -- and I don't think it's _too_ hard [to implement](https://jspecify.dev/tsttcpw.html#nullness-subtyping), even if the spec [resorts to handwaving](https://groups.google.com/d/msg/jspecify-dev/yPnkx_GSb0Q/hLgS_431AQAJ) a little.

Non-transitive subtyping is scary, but the alternative seems to be: "Subtyping is transitive, but you don't actually check subtyping anymore. Instead, every place where you _would_ have used subtyping, you use \<some other check\> instead."\[*\]That alternative sounds like the exact same thing, only under a different name :) It's scary either way, so I think we should be direct about it and say "subtyping is non-transitive."

\[*\] We'd have to use such a check for more than just "conversions.": For example, we would also use it to check type arguments. So I am wary of even saying the word "conversions": It may suggest "This is just like unchecked conversions \[and thus it's a solved problem\]" when in fact there's more to it.

**Comment by mernst** [comment]
I don't think that Jspecify should try to "conceptualize subtyping for legacy-nullness" (per the title of this issue).  Putting `@UnknownNullness` at a specific location in the type hierarchy of `@Nullable` and `@NonNull` implies a subtyping relationship that every tool must respect.  No such subtyping relationship exists, because `@UnknownNullness` may be treated differently by different tools.  Jspecify should state `@Nullable :> @NonNull`, but different tools may place `@UnknownNullness` at different locations within that type hierarchy.

Similarly, Jspecify should not dictate a tool's assignability rules.  Different tools will forbid/permit different assignments when `@UnknownNullness` is involved.

The  English description of `@UnknownNullness` can use type system concepts, so long as it does not dictate how a tool treats `@UnknownNullness`.  For instance, it might say that `@UnknownNullness` is wildcard-like in that a complete specification would give each type `@NonNull` or `@Nullable`, but in an incomplete specification (one that uses `@UnknownNullness`) the type's qualifier in Jspecify's type system is unknown.  But it would not say that tools must treat `@UnknownNullness` the same way that the Java type system treats wildcards (even though some particular tool might make that choice).

The discussion in this thread about putting `@UnknownNullness` in the type hierarchy or inventing new assignability rules might be useful for the implementation of one particular tool.  However, it is outside the scope of the Jspecify project proper.  I think that the Jspecify team should avoid such discussions, because they muddy the distinction between the definition of Jspecify and the behavior of a tool, and because they bias people toward one particular implementation choice that is not appropriate for all tools.

TL;DR: `@UnknownNullness` is outside the type hierarchy, from Jspecify's perspective.  Jspecify should not impose subtyping or assignability rules related to `@UnknownNullness`.

**Comment by cpovirk** [comment]
Additionally: I understand that `@NonNull <: @NullnessUnspecified <: @Nullable` is not sufficient to get all the properties that all tools might want. Still, do you see cases in which it gives results that a tool would consider _unsound_? I'm hoping that it's sound but incomplete.

Hmm, in asking this, I'm probably making a lot of assumptions about what `@NullnessUnspecified` (and for that matter, `@Nullable`) means. Under different assumptions, the thing that I just said is probably _not_ true.

**Comment by mernst** [comment]
> How would you feel about the idea that Jspecify supplies subtyping rules but tools can apply those rules when (and only when) they see fit?

What is the benefit of Jspecify taking a stand on subtyping of `@UnspecifiedNullness`, but permitting tools to ignore it and/or impose their own different subtyping?

**Comment by kevinb9n** [comment]
When searching for the elusive the line between what we should specify and what we should leave open to tools, I just want to opine again that our primary guiding star must always be, "if it influences how an API owner would choose to annotate their API, then we must specify it."

Can I annotate a return type as non-null if its supermethod returns unspecified? Can I annotate it as nullable if it already has overrides that may be unspecified?

That is a clear example of a question that needs definitive answers, and I've been seeing specifying subtyping rules as just a way to achieve that. But the same concern exists on a more subtle level too:

If I annotate this parameter type as nullable, how do I know I'm not causing a problem for any users no matter what they're passing in?

Essentially, every time I choose a parameter type, a part of that decision is always "and why is this not the next more general type, and why is this not the next more specific type?" For example, why did I settle on Collection, rather than Iterable or List?  I think I need to be able to do that for nullness too, and that's why it seems important that everyone agree on what the subtyping rules are.

**Comment by kevinb9n** [comment]
I should have mentioned that I do think we very often forget to think about that line on a daily basis, and it's right to call our attention to it so we can make sure we're respecting it.

I guess I can't really picture how tools could choose different subtyping rules unless they were also assuming different *meanings* for our annotations themselves. Or how they could do so without putting an API owner in the position of having to choose whether to make their X-tool users or their Y-tool users happy.

**Comment by mernst** [comment]
> Can I annotate a return type as non-null if its supermethod returns unspecified? Can I annotate it as nullable if it already has overrides that may return unspecified?

I'm not sure what question you are asking in "Can I".
 * If you are asking whether this is syntactically correct, then the answer is yes, because Jspecify does not forbid it.
 * If you are asking whether this causes a tool to issue a warning, that is a question that different tools may have different answers to.  Jspecify **cannot** and **must not** take a position on this question.

> I guess I can't really picture how tools could choose different subtyping rules unless they were also assuming different meanings for our annotations themselves. 

Yes, different tools may treat `@UnspecifiedNullness` differently.  I'm not sure what other annotations you are talking about when you say "our annotations", since this discussion is only about `@UnspecifiedNullness`.

**Comment by cpovirk** [comment]
> What is the benefit of Jspecify taking a stand on subtyping of `@UnspecifiedNullness`, but permitting tools to ignore it and/or impose their own different subtyping?

Fair: At this point, I am throwing ideas at the wall and seeing what sticks. If I'm not being outright self-contradictory, I'm getting very close. Still, I will try to wriggle out of this :)

In proposing that tools can ignore rules, I am mostly just acknowledging reality: I can write as many tens of thousands of words as I want, but ultimately tool owners get to decide what their tools do -- including, say, not recognizing Jspecify annotations at all, treating them [like the FindBugs `@Nullable`](http://findbugs.sourceforge.net/manual/annotations.html), or [checking only for contradictory constructions like `@Nullable @NonNull`](https://github.com/jspecify/jspecify/issues/92#issuecomment-578559822). I can _say_ that one interpretation or another is "required," but that distinction is not necessarily meaningful to end users. So I'm happy to push for weaker requirements if that makes this process less controversial.

But it may be that a weaker spec ends up _more_ controversial! Certainly there will be some controversy no matter what: After all, we're here because we assume that whatever "requirements" or "suggestions" we write will have _some_ effect. Maybe I am just playing semantic games? I don't feel that I have a well-thought out position of my own for this, and I surely haven't internalized your whole position, either.

For the purposes of our present discussions, I'd like to have _some_ term (whether it's "subtyping" or not) that we can use. With it, we can discuss behavior that we expect some tools to find useful but that we acknowledge that not all tools will implement. I think that providing a guide to such "potentially useful behavior" is a noble goal for us, even if it's "not a spec."

Yet, at the same time, I am poking you with question about whether `@NonNull <: @NullnessUnspecified <: @Nullable` is sound, presumably under the assumption that you and other tool authors might adopt it :) I do want for our "suggestions" to be palatable to tool authors. Not to mention that I want to be unsound only intentionally, rather than by accident. As much as possible, I'd like for tool authors to feel that they need only layer knowledge _on top of_ ours: "Jspecify doesn't know that this this expression's type is a subtype of `@NonNull Object`, but _I_ know that someone just performed a null check on it, so...."

**Comment by cpovirk** [comment]
@mernst , I'm sorry if you've stated this before and I've forgotten it, but do you have an idea in mind of where you would draw the line between what to specify and what not to? I just linked to [one thing you said](https://github.com/jspecify/jspecify/issues/92#issuecomment-578559822), but I don't want to make assumptions about what you meant by "perhaps a few other weak properties."

For example, in [endorsing `@Nullable :> @NonNull`](https://github.com/jspecify/jspecify/issues/33#issuecomment-636123897), it seems like you **might** consider the following test to be acceptable:

```
Object foo(@Nullable Object p) {
  return p; // type mismatch
}
```

In terms of what to check, I'm thinking a little in terms of [the Subtyping Checker](https://checkerframework.org/manual/#subtyping-checker) here. (You would know better than I if that comparison is in any way illuminating or even fair.)

But it may be that even that is too far: If we're not prescribing tool behavior, then what does it mean to "test" what happens when there's a mismatch?

**Comment by kevinb9n** [comment]
> > Can I annotate a return type as non-null if its supermethod returns unspecified? Can I annotate it as nullable if it already has overrides that may return unspecified?
> 
> I'm not sure what question you are asking in "Can I".
> 
> * If you are asking whether this is syntactically correct, then the answer is yes, because Jspecify does not forbid it.

I think it should be more than "I searched everything and couldn't find anything saying not to, so I guess I can". It should be that any person who understands the meanings of the annotations correctly will have *positive* knowledge that of course this scenario above is okay. I brought the word "subtyping" into all this as a way to provide that positive knowledge. When a person knows that a subtyping relationship exists then they can relax in the knowledge that everything true of the parent is true of the child.

I do get the sense though that you and I are often arguing about something different from what we really seem to be arguing about, but I can't always put my finger on what our _real_ disagreement is.


> * If you are asking whether this causes a tool to issue a warning, that is a question that different tools may have different answers to.  Jspecify **cannot** and **must not** take a position on this question.

I hope I can explain this right. I've tried a number of times.

I really and truly don't care what specific warnings tools choose to emit.

I really and truly do want our annotations to have clear and precise meanings. The result of that is that certain combinations of annotations will be outright logical contradictions. I want our specs to be clear on those combinations being invalid. I don't care what a tool does or doesn't decide to do about it.

I *am* concerned about tools deciding that one of our annotations carries some meaning that is more and/or less than what we specified, and acting accordingly to that interpretation. Not that I have any illusions that we can prevent it, but I think it is decidedly a bad outcome, and if our decisions pushed the tool owner to that route then we would have to accept some of the blame. The poster child for this is of course `@NullnessUnspecified`, because it represents the absolute lack of *any* information about the intended nullness of the type. If a tool interprets that annotation as, "we think what they're really trying to say is...", well, again, I just think that's a bad outcome.


> Yes, different tools may treat `@UnspecifiedNullness` differently. I'm not sure what other annotations you are talking about when you say "our annotations", since this discussion is only about `@UnspecifiedNullness`.

And I think this glosses over the distinction I've been struggling to make (perhaps because it is just too subtle for its own good). The outcome I hope for is that any two tools are going to agree that the annotations *mean* the same thing (convey the same information about the API owner's intentions), they may just want to take different *actions* based on the same information. I can't tell if you understand it the same way and just keep seeing us as trying to dictate the actions, or if this is really what you're objecting to itself.

**Comment by kevinb9n** [comment]
(A lot of this is the issue that #73 was initially created to discuss, though who knows in what directions the discussion has gone since :-))

**Comment by kevin1e100** [comment]
Two questions, @mernst:
1. What's the value of not specifying a (partial or total) ordering of the three qualifiers?  I guess what I'm asking is, can you give some examples of what only specifying "non-null < nullable", and saying nothing about how "unspecified" relates, would allow tools to do that specifying "non-null <= unspecified <= nullable" would prevent?

1. If we define "unspecified" as "it may be non-null or it may be nullable, we don't know", it appears:
   * non-null implies nullable
   * non-null implies unspecified
   * unspecified implies nullable
  Seemingly these implications hold regardless of whether we prescribe them, just following the definition.  Even if we define `unspecified := \exists x. (x = non-null or nullable)`, that is, even if we defined "unspecified" as derived from other qualifiers, these implications seem to hold.  If that's the case then it seems "harmless" to codify something that appears true anyways.  What am I missing?

That said I agree that defining subtyping or assignability could be overreaching, not least since assignments happen in code, which we consider out of scope.  There's however one area where we seemingly should say _something_, and that is the area of what constitutes valid overriding.  That is, it seems to me it would be useful to introduce some lowest common denominator where all tools are required to consider some declarations `class C extends B implements Foo, Bar, ... { ... }` as "well formed" so they can interpret them consistently (and I'm talking about declarations only here, not method implementation code).  For instance, consider:

```
// no annotations
interface ListLike<T> {
  T get(int index);
  boolean contains(Object elem);
}
// annotated
class MyList implements ListLike<@NonNull String> {
  @Override T get(int index) {...}
  @Override boolean contains(@Nullable Object elem) {...}
}
```

Given `ListLike`, it seems helpful to prescribe whether `MyList` should be considered a well-formed declaration, that defines a class that implements `ListLike<@NonNull String>`, with methods that indeed override `ListLike<@NonNull String>`'s methods.  If we don't do that, but the author of `MyList` happens to use a tool that accepted `MyList`, a second tool might reject `MyList` or might refuse to process code that depends on `MyList` or even refuse to read in `MyList`'s bytecode.  We can argue over exactly what that bar should be, and whether it should apply only to bytecode or also source code, but it seems useful to define some bar.  And because most code today is unannotated, it seems important for that bar to include some rules for handling unspecified nullness.

It seems we may get away with simply saying something like, "it's always allowed to use `@UnspecifiedNullness` (implicitly in unannotated code, or explicitly) where a supertype used `@NonNull` or `@Nullable`, and vice versa", which doesn't require an ordering of unspecified vs. other qualifiers.  We could still specify co-, contra-, and in-variant handling of `@NonNull` vs `@Nullable` as appropriate, just not for `@UnspecifiedNullness`.  We can certainly allow tools to raise warnings when seeing changes to or from `@UnspecifiedNullness` in source code, maybe even in bytecode, but it does seem useful to define "enough" for tools to still interpret `MyList` consistently, even if some of them raise warnings on `MyList`'s declaration, so that references to `MyList` in other code can be handled.  I believe this level of specification would be consistent with not including "unspecified" in any kind of ordering as you requested.  Is that what you had in mind?

I personally would like to take one or more of the following steps (still talking only about declarations and overriding here, not code, and specifically overriding that changes `@UnspecifiedNullness` to one of the other 2 qualifiers), in order of preference:
1. some form of `@SuppressWarnings`-like annotation should suppress warnings related to overriding that changes qualifiers to and from `@UnspecifiedNullness` in the annotated scope.  This would make it so users have a way to make any tool accept the overriding they want.
1. if we draw a distinction between warnings, where tools print something but continue, and errors, where tools stop, tools should never "error" when they see these situations in bytecode (e.g., tools would be required to accept `MyList`'s bytecode: they can print warnings but should continue on).
1. changes to and from `@UnspecifiedNullness` that are consistent with the proposed qualifier ordering (non-null <= unspecified <= nullable) and allowed per co- or contra-variance rules shouldn't cause warnings, or at least shouldn't cause errors (because they are seemingly "safe" as discussed above).

Now only the last one requires defining some kind of ordering of all 3 qualifiers.  I do think it's a useful rule that would make interop between different tools more seamless.  But I guess in all of this I've talked myself into thinking that we can get pretty far even without defining this ordering.  Thoughts?

**Comment by kevin1e100** [comment]
One other thing to note is that not ordering "unspecified" seemingly has both perils and advantages in this weird case, which I implicitly referenced in `ListLike<T>` in my last post:
```
interface Foo<T extends @NullnessUnspecified Object>
```
It's always been a bit vexing what this is supposed to mean.
* by not relating "unspecified" to other qualifiers, we're somewhat side-stepping that question, and in particular that would allow tools to raise warnings on `Foo<@NonNull String>`, which multiple folks have considered desirable.
* by stating that non-null <= unspecified we're apparently saying that `Foo<@NonNull String>` should be considered valid, which I personally think is fine, but it's then seemingly inconsistent that `Foo<@Nullable String>` can cause warnings.

**Comment by kevinb9n** [comment]
> Two questions, @mernst:
> 
>      Seemingly these implications hold regardless of whether we prescribe them, just following the definition.

Yes, thank you for that clarity, this is my view too.


> That said I agree that defining subtyping or assignability could be overreaching, not least since assignments happen in code, which we consider out of scope.

Just to renew my objection to this: a conscientious API owner decides what type a parameter should have based on what kinds of actual parameter expressions should be assignable to it.

I think "APIs, not implementation code" has been a slight oversimplification, and we might as well just use this better oversimplification instead:

* APIs: in scope
* Implementation code that is directly interacting with an API: in scope
* Other implementation code, out of scope

> I personally would like to take one or more of the following steps (still talking only about declarations and overriding here, not code, and specifically overriding that changes `@UnspecifiedNullness` to one of the other 2 qualifiers), in order of preference:
> 
> 1. some form of `@SuppressWarnings`-like annotation should suppress warnings related to overriding that changes qualifiers to and from `@UnspecifiedNullness` in the annotated scope.  This would make it so users have a way to make any tool accept the overriding they want.

(aside: This should be @SW itself, since one of the reasons the qualifier is getting overridden to unspec may be that the library is not taking on our dependency yet.)


> 2. if we draw a distinction between warnings, where tools print something but continue, and errors, where tools stop, tools should never "error" when they see these situations in bytecode (e.g., tools would be required to accept `MyList`'s bytecode: they can print warnings but should continue on).

To head off Mike's probable objection :-), we *don't* need to take stances on what tools print.

Once we provide precise enough meanings for our annotations, then we can point out various situations that are categorically invalid ("error") *as direct consequences* of those meanings, and situations that are less than provably safe ("warning") as well. And we should. We certainly mustn't *avoid* doing so out of fear that it could sound like dictating tool behavior.

> But I guess in all of this I've talked myself into thinking that we can get pretty far even without defining this ordering. Thoughts?

I'm still in sync with your opener: that the ordering is nothing but a logical consequence of the definitions. One reaction to that is "so we don't need to specify it", but mine is "so it would be silly of us to pretend that we're not specifying it." We are, just by consequence rather than fiat.

**Comment by kevinb9n** [comment]
Btw, the more repeatedly and strenuously I assert a *position*, the more I welcome comments like "and here are some consequences of that position you might not be accounting for." That, in my view, is how this all works.

**Comment by kevinb9n** [comment]
I'm convicting this issue of the crime of being extremely long and outdated. Anyone may file a new issue for any particular aspect touched on here that feels in need of resolution.
